test_application/src/main.rs

```diff
@@ -0,0 +1,272 @@
+    start_subprotocol(&next_action, SubProtocol::StartPresign)?;
+    route_worker_messages(&workers_message, &workers_rx, num_workers)?;
+    info!("Finished presign.");
+
+    info!("Starting sign.");
+    start_subprotocol(&next_action, SubProtocol::Sign)?;
+    route_worker_messages(&workers_message, &workers_rx, num_workers)?;
+    info!("Finished sign.");
+
+    Ok(())
+}
+
+fn start_subprotocol(
+    next_action: &NextActionChannels,
+    sub_protocol: SubProtocol,
+) -> anyhow::Result<()> {
+    debug!("Starting sub-protocol: {:?}", sub_protocol);
+    let rng = &mut thread_rng();
+    let sid = Identifier::random(rng);
+
+    for worker in next_action {
+        worker.send((sub_protocol, sid))?;
+    }
+
+    Ok(())
+}
+
+fn route_worker_messages(
+    send_to_worker: &WorkerChannelsMessage,
+    from_workers: &Receiver<MessageFromWorker>,
+    num_workers: usize,
+) -> anyhow::Result<()> {
+    // # of workers who finished the current sub_protocol.
+    let mut sub_protocol_ended = 0;
+
+    for message in from_workers {
+        debug!("Received messages from a worker.");
+        match message {
+            MessageFromWorker::FinishedRound(messages) => {
+                if messages.is_empty() {
+                    // TODO: Empty messages are awkward...
+                    trace!("Received empty messages...");
+                }
+                for m in messages {
+                    trace!("Routing message: {:?}", m);
+                    let recipient = m.to();
+                    send_to_worker.get(&recipient).unwrap().send(m)?;
+                }
+            }
+            MessageFromWorker::SubProtocolEnded => {
+                trace!("Worker finished sub-protocol.");
+                sub_protocol_ended += 1;
+                if sub_protocol_ended == num_workers {
+                    debug!("All workers finished sub-protocol. Entire sub-protocol ended.");
+                    break;
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+/// Workers have a single incoming and outgoing channel (connected to the
+/// router).
+#[instrument(skip_all)]
+fn participant_worker(
+    config: ParticipantConfig,
+    next_action: Receiver<(SubProtocol, Identifier)>,
+    other_workers: Receiver<Message>,
+    outgoing: Sender<MessageFromWorker>,
+) -> anyhow::Result<()> {
+    info!("Worker thread started.");
+    let mut public_key_shares: Option<Vec<KeySharePublic>> = None;
+    let mut private_key_share: Option<KeySharePrivate> = None;
+    let mut aux_info_public: Option<Vec<AuxInfoPublic>> = None;
+    let mut aux_info_private: Option<AuxInfoPrivate> = None;
+    let mut presign_record: Option<PresignRecord> = None;
+
+    for (sub_protocol, sid) in next_action {
+        match sub_protocol {
+            SubProtocol::StartKeyGen => {
+                // TODO: Whats the point of making a stand-alone `KeygenParticipant`
+                // let key_gen = KeygenParticipant::new(config.id.clone(),
```

Comment:
I'm not sure if I like the concept of a `config`, at least the name? But this is a minor nitpick.